EIPS/eip-9999.md

```diff
@@ -0,0 +1,447 @@
+
+# Storage change: [tx_index, new_value]
+StorageChange = [TxIndex, StorageValue]
+
+# Balance change: [tx_index, post_balance]
+BalanceChange = [TxIndex, Balance]
+
+# Nonce change: [tx_index, new_nonce]
+NonceChange = [TxIndex, Nonce]
+
+# Code change: [tx_index, new_code]
+CodeChange = [TxIndex, CodeData]
+
+# All changes to a single storage slot
+SlotChanges = [StorageKey, List[StorageChange]]
+
+# All changes for a single account within the chunk
+AccountChanges = [
+    Address,                    # address
+    List[SlotChanges],         # storage_changes
+    List[StorageKey],          # storage_reads (read-only)
+    List[BalanceChange],       # balance_changes
+    List[NonceChange],         # nonce_changes
+    List[CodeChange]           # code_changes
+]
+
+# Chunk Access List: All account changes in the chunk
+ChunkAccessList = List[AccountChanges]
+```
+
+CALs are incremental. This means that changes and storage reads from earlier chunks are not included in CALs of later chunks.
+
+#### Network Sidecars
+
+Chunks and CALs propagate as separate sidecars. They can arrive in any order, including before the beacon block.
+
+```python
+# Chunk propagated as sidecar
+class ExecutionChunkSidecar(Container):
+    root: Root                          # The beacon block root this chunk belongs to
+    chunk: ExecutionChunk               # The chunk
+    chunk_signature: BLSSignature       # The builder's signature
+
+# CAL propagated as separate sidecar
+class ChunkAccessListSidecar(Container):
+    root: Root                          # The beacon block root this chunk belongs to
+    chunk_index: ChunkIndex             # The index of the chunk in the block
+    chunk_access_list: ChunkAccessList  # The chunk access list
+    cal_signature: BLSSignature         # The builder's signature
+```
+
+### Chunk Construction Rules
+
+Block producers MUST follow these rules when creating chunks:
+
+1. **Gas Limits**: Each chunk MUST NOT exceed `CHUNK_GAS_LIMIT` (16,777,216 gas)
+2. **Invalid Merge**: The combined gas used by two consecutive chunks MUST exceed `CHUNK_GAS_LIMIT` (it should be impossible to simply merge two consecutive chunks and make a single valid chunk)
+3. **Transaction Integrity**: Transactions MUST NOT be split across chunks
+4. **Withdrawal Placement**: Withdrawals MUST only appear in the final chunk
+5. **Maximum Chunks**: A block MUST contain at most `MAX_CHUNKS_PER_BLOCK` chunks
+6. **Ordering**: Chunks MUST be numbered sequentially from 0 to N-1
+
+### Execution Semantics
+
+#### Streaming Chunk Execution
+
+The execution layer processes chunks independently, where each chunk applies all prior CALs to the parent state:
+
+```python
+def el_validate_chunk(header, chunk, cal, prior_cals):
+    """EL executes chunk using CALs 0..N-1 applied to parent state"""
+    chunk_index = chunk.chunk_header.index
+    
+    # Validate CAL availability
+    if len(prior_cals) != chunk_index:
```

Comment:
So the CALs can arrive in any order, if I would add them to the list of CALs which I have received, then `prior_cals` does not necessarily have the CALs which I need (there might be gaps). So for a chunk, what we would need is either have all previous chunks executed, or we need all previous CALs so we can build the pre-state. If any of this is true we can start executing (we do not need the CAL of this chunk before applying this). When this is done, we verify that the CAL we produced equals the CAL received.

Reply:
I might also have missed some point here though 😅